switch_model/tools/graphing/main.py

```diff
@@ -0,0 +1,466 @@
+
+original_working_dir = os.getcwd()
+
+
+class GraphDataFolder:
+    OUTPUTS = "outputs"
+    INPUTS = "inputs"
+
+
+class Scenario:
+    """
+    Stores the information related to a scenario such as the scenario name (used while graphing)
+    and the scenario path.
+
+    Also allows doing:
+
+    with scenario:
+        # some operation
+
+    Here, some operation will be run as if the working directory were the directory of the scenario
+    """
+    root_path = os.getcwd()
+
+    def __init__(self, rel_path, name):
+        self.path = os.path.join(Scenario.root_path, rel_path)
+        self.name = name
+
+        if not os.path.isdir(self.path):
+            raise Exception(f"Directory does not exist: {self.path}")
+
+    def __enter__(self):
+        os.chdir(self.path)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        os.chdir(Scenario.root_path)
+
+
+class GraphData:
+    """
+    Object that stores and handles loading csv into dataframes data for different scenarios.
+    """
+
+    def __init__(self, scenarios: List[Scenario]):
+        self.scenarios: List[Scenario] = scenarios
+
+        # Here we store a mapping of csv file names to their dataframes.
+        # Each dataframe has a column called 'scenario' that specifies which scenario
+        # a given row belongs to
+        self.dfs: Dict[str, pd.DataFrame] = {}
+
+        # Check that the scenario names are unique
+        all_names = list(map(lambda s: s.name, scenarios))
+        if len(all_names) > len(set(all_names)):  # set() drops duplicates, so if not unique len() will be less
+            raise Exception("Scenario names are not unique.")
+
+        # Disables warnings that will occur since we are constantly returning only a slice of our master dataframe
+        pd.options.mode.chained_assignment = None
+
+    def _load_dataframe(self, csv, folder):
+        """Loads the dataframe into self.dfs[csv]"""
+        df_all_scenarios: List[pd.DataFrame] = []
+        for i, scenario in enumerate(self.scenarios):
+            df = pd.read_csv(os.path.join(scenario.path, folder, csv + ".csv"), index_col=False)
+            df['scenario_name'] = scenario.name
+            df['scenario_index'] = i
+            df_all_scenarios.append(df)
+
+        self.dfs[csv] = pd.concat(df_all_scenarios)
+
+    def get_dataframe(self, scenario_index, csv, folder=GraphDataFolder.OUTPUTS):
+        """Return a dataframe filtered by the scenario_name"""
+        if csv not in self.dfs:
+            self._load_dataframe(csv, folder)
+
+        df = self.dfs[csv]
+        return df[df['scenario_index'] == scenario_index]
+
+    def get_dataframe_all_scenarios(self, csv, folder=GraphDataFolder.OUTPUTS):
+        """Fetch the dataframe containing all the scenarios"""
+        if csv not in self.dfs:
+            self._load_dataframe(csv, folder)
+
+        return self.dfs[csv].copy()  # We return a copy so the source isn't modified
```

Comment:
The purpose of this class seems to be an enhancement of the Scenario class (load all the respective data frames per scenario) and also an easy way to get the data for multiple scenarios. I do not see the benefit of having them in different classes. Could you please show me (high-level) how do you envision the use of both classes? Because this class seems more oriented to the compare functionality rather than standard graphing.

Reply:
Based on your feedback I realized this is a bit overly complex. I merge the functionality direcrlt into `GraphTools` and that reduced the amount of code quite a bit. To explain nonetheless, this functionality is quite different from the `Scenario` class. Every `Scenario` holds information about a single scenario, while this class (which is now in `GraphTools`) holds the data across all scenarios in a single dataframe (per csv file).